#ifndef TCPSOCKET_H
#define TCPSOCKET_H
#include <QObject>

class OrderInterface;
class OrderFrameBuffer;

class QString;
class QTcpSocket;

/*
该类应用于与服务器的网络交互
a.请求链接
b.控制与kinect设备的链接
c.断开链接
该类所对应的接口类为 OrderInterface
需要先调用registerGUIClass()后才可以调用start()以开启线程读取缓存
*/
class OrderSocket : public QObject
{
    Q_OBJECT
public:
    explicit OrderSocket(QString adress="127.0.0.1", unsigned int port= 7892, QObject *parent = nullptr);
    ~OrderSocket();

private:
    QString m_strIPAdress;
    unsigned int m_uPort;
    QTcpSocket *m_pTcpSocket;

private:
    bool m_bConnected = false;
private slots:
    inline void slot_setConnected();
    inline void slot_setDisConnected();
/***********************************************************************************/
/*与GUI通信所用接口*/
public:
    void registerGUIClass(OrderInterface*  gui)
    {
        this->m_pGUI = gui;
    };
private:
    OrderInterface *m_pGUI = nullptr;
/***********************************************************************************/

/***********************************************************************************/
/*发送指令到服务器的相关函数*/
private:
    OrderFrameBuffer *m_pSendFrameBuffer = nullptr;

private:
    bool writeBufferToServer() const;
    bool writeBufferToServer(const OrderFrameBuffer & buffer) const;

public slots:
    bool slot_requireConnect();
    bool slot_exitConnect();
    bool slot_requireDevices();
    bool slot_startRequire(QString deviceName, unsigned int dataType);
    bool slot_endConnect();
/***********************************************************************************/

/***********************************************************************************/
/*与线程相关的变量以及函数*/
protected:
    QByteArray m_receiveBuffer;

    bool m_bNotHasHead = true;
    OrderFrameBuffer *m_pReceiveFrameBuffer = nullptr;
private:
    void analysisReceiveByteArrayBuffer();
    void analysisReceiveFrameBuffer(const OrderFrameBuffer &buffer);

private slots:
    void slot_readDataFromServer();
/***********************************************************************************/
};

#endif // TCPSOCKET_H
